passport.go

```diff
@@ -92,43 +97,83 @@ func (v *Verifier) VerifyProof(proof zkptypes.ZKProof, options ...VerifyOption)
 		// Required fields to validate
-		"pub_signals/nullifier":       val.Validate(zkProof.PubSignals[PubSignalNullifier], val.Required),
-		"pub_signals/selector":        val.Validate(zkProof.PubSignals[pubSignalSelector], val.Required, val.In(proofSelectorValue)),
-		"pub_signals/expiration_date": val.Validate(zkProof.PubSignals[pubSignalExpirationDate], val.Required, afterDate(time.Now().UTC())),
-		"pub_signals/id_state_hash":   err,
+		"pub_signals/nullifier":                   val.Validate(zkProof.PubSignals[Nullifier], val.Required),
+		"pub_signals/selector":                    val.Validate(zkProof.PubSignals[Selector], val.Required, val.In(proofSelectorValue)),
+		"pub_signals/expiration_date_lower_bound": val.Validate(zkProof.PubSignals[ExpirationDateLowerBound], val.Required, afterDate(time.Now().UTC())),
+		"pub_signals/id_state_hash":               err,
 
 		// Configurable fields
-		"pub_signals/event_id": val.Validate(zkProof.PubSignals[pubSignalEventID], val.When(
+		"pub_signals/event_id": val.Validate(zkProof.PubSignals[EventID], val.When(
 			!val.IsEmpty(v.opts.eventID),
 			val.Required,
 			val.In(v.opts.eventID))),
-		"pub_signals/birth_date": val.Validate(zkProof.PubSignals[pubSignalBirthDate], val.When(
+		"pub_signals/birth_date_upper_bound": val.Validate(zkProof.PubSignals[BirthdateUpperBound], val.When(
 			!val.IsEmpty(v.opts.age),
 			val.Required,
 			beforeDate(v.opts.age),
 		)),
-		"pub_signals/citizenship": val.Validate(decodeInt(zkProof.PubSignals[pubSignalCitizenship]), val.When(
+		"pub_signals/citizenship": val.Validate(decodeInt(zkProof.PubSignals[Citizenship]), val.When(
 			!val.IsEmpty(v.opts.citizenships),
 			val.Required,
 			val.In(v.opts.citizenships...),
 		)),
-		"pub_signals/event_data": val.Validate(zkProof.PubSignals[pubSignalEventData], val.When(
-			!val.IsEmpty(v.opts.address),
+		"pub_signals/event_data": val.Validate(zkProof.PubSignals[EventData], val.When(
+			!val.IsEmpty(v.opts.eventData),
 			val.Required,
-			matchesAddress(v.opts.address),
+			val.In(v.opts.eventData),
 		)),
 	}.Filter()
 }
+
+func (v *Verifier) validateIdentitiesInputs(signals []string) error {
+	counterUpperBound, err := strconv.ParseInt(signals[IdentityCounterUpperBound], 10, 64)
+	if err != nil {
+		return val.Errors{
+			"pub_signals/identity_counter_upperbound": fmt.Errorf("parsing integer: %w", err),
+		}.Filter()
+	}
+
+	counterErr := val.Errors{
+		"pub_signals/identity_counter_upperbound": val.Validate(counterUpperBound, val.When(
+			v.opts.maxIdentitiesCount != -1,
+			val.Required,
+			val.Max(v.opts.maxIdentitiesCount),
+		)),
+	}.Filter()
+
+	timestampErr := val.Errors{
+		"pub_signals/timestamp_upperbound": val.Validate(signals[TimestampUpperBound], val.When(
+			!val.IsEmpty(v.opts.lastIdentityCreationTimestamp),
+			val.Required,
+			beforeDate(v.opts.lastIdentityCreationTimestamp),
+		)),
+	}.Filter()
+
+	if counterErr != nil && timestampErr != nil {
+		return errors.Join(counterErr, timestampErr)
```

Comment:
Still a crutch, you lose validation.Errors. If we can't avoid merging, let's merge it with iteration over a map (Errors is map[string]error)